src/stream.rs

```diff
@@ -435,11 +435,16 @@ pub(crate) fn connect_host(
 
             let mut proxy_response = Vec::new();
 
+            let mut buf = [0u8; 1];
             loop {
-                let mut buf = vec![0; 256];
                 let total = stream.read(&mut buf)?;
-                proxy_response.append(&mut buf);
-                if total < 256 {
+                if total == 0 {
+                    // no more data
+                    break;
+                }
+
+                proxy_response.push(buf[0]);
```

Comment:
While this code is a big improvement on the previous (which looks broken), there's one problem here.

Perhaps a very slim chance, but a misbehaving proxy could potentially just keep writing anything but `\r\n\r\n` and cause a denial of service attack by exhausting the available RAM by this `proxy_response` ever growing.

In the "normal" header parsing this is capped by `MAX_HEADER_COUNT` and `MAX_HEADER_SIZE`. We need to do _something_ here as well. Ideally it would use the same logic as the main header parsing (maybe we can break that out into a separate subfunction and reuse in both call sites?)

Or we need to cap the proxy_response at something like 1MB.

Reply:
Ok, will look into that.

Reply:
I've looked into it and we could actually just use the `Response` struct for parsing the response.
The easiest way to do this would be to pass a Stream to Response::do_from_stream. It requires cloning the TcpStream though.
I'm not sure if I'm missing something regarding Stream and its handling of the underlying TcpStream, especially in regards to dropping.

The response can then be used to check the status code instead of parsing the string again in verify_response.

Would that be acceptable for you or would you prefer a different solution?